Source/Core/InputCommon/ControllerInterface/WGInput/WGInput.h

```diff
@@ -0,0 +1,222 @@
+  };
+
+  // A button with one of the "labels" that WGI provides:
+  class NamedButton : public Button
+  {
+  public:
+    NamedButton(const boolean& _button, const char* _name) : Button(_button), m_name(_name) {}
+    std::string GetName() const override;
+
+  private:
+    const char* const m_name;
+  };
+
+  // A button with no label so we name it by its index:
+  class IndexedButton : public Button
+  {
+  public:
+    IndexedButton(const boolean& _button, u32 _index) : Button(_button), m_index(_index) {}
+    std::string GetName() const override;
+
+  private:
+    u32 m_index;
+  };
+
+  class Axis : public Core::Device::Input
+  {
+  public:
+    Axis(const double& _axis, double _base, double _range)
+        : m_base(_base), m_range(_range), m_axis(_axis)
+    {
+    }
+
+    ControlState GetState() const override;
+
+  protected:
+    const double m_base, m_range;
+
+  private:
+    const double& m_axis;
+  };
+
+  class NamedAxis : public Axis
+  {
+  public:
+    NamedAxis(const double& _axis, double _base, double _range, const char* _name)
+        : Axis(_axis, _base, _range), m_name(_name)
+    {
+    }
+    std::string GetName() const override;
+
+  private:
+    const char* const m_name;
+  };
+
+  class IndexedAxis : public Axis
+  {
+  public:
+    IndexedAxis(const double& _axis, double _base, double _range, u32 _index)
+        : Axis(_axis, _base, _range), m_index(_index)
+    {
+    }
+    std::string GetName() const override;
+
+  private:
+    const u32 m_index;
+  };
+
+  class NamedTrigger : public Axis
+  {
+  public:
+    NamedTrigger(const double& _axis, const char* _name) : Axis(_axis, 0.0, 1.0), m_name(_name) {}
+    std::string GetName() const override;
+
+  private:
+    const char* const m_name;
+  };
+
+  class Switch : public Core::Device::Input
+  {
+  public:
+    Switch(const WGI::GameControllerSwitchPosition& _switch, u32 _index,
+           WGI::GameControllerSwitchPosition _direction)
+        : m_switch(_switch), m_index(_index), m_direction(_direction)
+    {
+    }
+    std::string GetName() const override;
+    ControlState GetState() const override;
+
+  private:
+    const WGI::GameControllerSwitchPosition& m_switch;
+    const u32 m_index;
+    const WGI::GameControllerSwitchPosition m_direction;
+  };
+
+protected:
+  void PopulateButtons(ComPtr<WGI::IRawGameController> controller, std::vector<boolean>& buttons);
+
+public:
+  Device(const std::string& _name) : m_name(_name) {}
+
+  std::string GetName() const override;
+  std::string GetSource() const override;
+
+private:
+  const std::string m_name;
+};
+
+class Gamepad : public Device
+{
+public:
+  Gamepad(ComPtr<WGI::IRawGameController> _raw_controller, WGI::IGamepad* _gamepad,
+          const std::string& _name);
+
+  void UpdateInput() override;
+  void UpdateMotors();
+
+private:
+  class NamedMotor : public Core::Device::Output
+  {
+  public:
+    NamedMotor(double& _motor, const char* _name, Gamepad& _parent)
```

Comment:
Please make these non-const refs pointers. That way it's clearer at invocation time that the function keeps a mutable reference to its arguments.